web/src/core/Crust/Plugins/PluginFrame/SafeIFrame/hooks.ts

```diff
@@ -0,0 +1,169 @@
+    const cb = (ev: MessageEvent<any>) => {
+      if (!iFrameRef.current || ev.source !== iFrameRef.current.contentWindow) return;
+      if (ev.data?.[autoResizeMessageKey]) {
+        const { width, height } = ev.data[autoResizeMessageKey];
+        if (typeof width !== "number" || typeof height !== "number") return;
+        setIFrameSize([width + "px", height + "px"]);
+        onAutoResized?.();
+      } else {
+        onMessage?.(ev.data);
+      }
+    };
+    window.addEventListener("message", cb);
+    return () => {
+      window.removeEventListener("message", cb);
+    };
+  }, [autoResize, autoResizeMessageKey, onMessage, onAutoResized]);
+
+  const onIframeLoad = useCallback(() => {
+    loaded.current = true;
+    onLoad?.();
+  }, [onLoad]);
+
+  const props = useMemo<IframeHTMLAttributes<HTMLIFrameElement>>(
+    () => ({
+      ...iFrameProps,
+      style: {
+        display: visible ? "block" : "none",
+        width: visible
+          ? !autoResize || autoResize == "height-only"
+            ? "100%"
+            : iFrameSize?.[0]
+          : "0px",
+        height: visible
+          ? !autoResize || autoResize == "width-only"
+            ? "100%"
+            : iFrameSize?.[1]
+          : "0px",
+        ...iFrameProps?.style,
+      },
+    }),
+    [autoResize, iFrameProps, iFrameSize, visible],
+  );
+
+  useEffect(() => {
+    const handleBlur = () => {
+      if (iFrameRef.current && iFrameRef.current === document.activeElement) {
+        onClick?.();
+      }
+    };
+    window.addEventListener("blur", handleBlur);
+    return () => {
+      window.removeEventListener("blur", handleBlur);
+    };
+  }, [onClick]);
+
+  useEffect(() => {
+    const w = typeof width === "number" ? width + "px" : width;
+    const h = typeof height === "number" ? height + "px" : height;
+    setIFrameSize(w || h ? [w, h] : undefined);
+  }, [width, height]);
+
+  const autoResizeScript = useMemo(() => {
+    return `<script id="_reearth_resize">
+      if ("ResizeObserver" in window) {         
+        new window.ResizeObserver(entries => {
+          const win = document.defaultView;
+          const html = document.body.parentElement;
+          const st = win.getComputedStyle(html, "");
+          horizontalMargin = parseInt(st.getPropertyValue("margin-left"), 10) + parseInt(st.getPropertyValue("margin-right"), 10);
+          verticalMargin = parseInt(st.getPropertyValue("margin-top"), 10) + parseInt(st.getPropertyValue("margin-bottom"), 10);
+          const width = html.offsetWidth + horizontalMargin;
+          const height = html.offsetHeight + verticalMargin;
+          if(parent){
+            parent.postMessage({
+              [${JSON.stringify(autoResizeMessageKey)}]: { width, height }
+            }, "*")
+          }
+        }).observe(document.body.parentElement);
+      }
+    </script>`;
+  }, [autoResizeMessageKey]);
+
+  const srcDoc = useMemo(() => {
+    return html ? `${html}${autoResizeScript}` : "";
```

Comment:
If html is like the following, does script tag work?
```html
<DOCTYPE!>
<html>
  <body>
    <p>Hello World</p>
  </body>
</html>
```

Reply:
Emmm it does not match W3C spec but most browser can still work with this. 
I updated the code to move the auto resize script into body if there's a end tag of body.